src/arfs_entities.ts

```diff
@@ -1,25 +1,63 @@
 import { FolderHierarchy } from './folderHierarchy';
 import {
 	ByteCount,
 	CipherIV,
 	DataContentType,
 	DriveID,
-	EntityID,
+	AnyEntityID,
 	FileID,
 	FolderID,
 	TransactionID,
 	UnixTime
 } from './types';
 
+// The primary ArFS entity that all other entities inherit from.
+export class ArFSEntity {
+	appName: string; // The app that has submitted this entity.  Should not be longer than 64 characters.  eg. ArDrive-Web
+	appVersion: string; // The app version that has submitted this entity.  Must not be longer than 8 digits, numbers only. eg. 0.1.14
+	arFS: string; // The version of Arweave File System that is used for this entity.  Must not be longer than 4 digits. eg 0.11
+	contentType: string; // the mime type of the file uploaded.  in the case of drives and folders, it is always a JSON file.  Public drive/folders must use "application/json" and priate drives use "application/octet-stream" since this data is encrypted.
+	driveId: DriveID; // the unique drive identifier, created with uuidv4 https://www.npmjs.com/package/uuidv4 eg. 41800747-a852-4dc9-9078-6c20f85c0f3a
+	entityType: string; // the type of ArFS entity this is.  this can only be set to "drive", "folder", "file"
+	name: string; // user defined entity name, cannot be longer than 64 characters.  This is stored in the JSON file that is uploaded along with the drive/folder/file metadata transaction
+	txId: string; // the arweave transaction id for this entity. 43 numbers/letters eg. 1xRhN90Mu5mEgyyrmnzKgZP0y3aK8AwSucwlCOAwsaI
+	unixTime: number; // seconds since unix epoch, taken at the time of upload, 10 numbers eg. 1620068042
+
+	constructor(
+		appName: string,
+		appVersion: string,
+		arFS: string,
+		contentType: string,
+		driveId: DriveID,
+		entityType: string,
+		name: string,
+		txId: string,
+		unixTime: number
+	) {
+		this.appName = appName;
+		this.appVersion = appVersion;
+		this.arFS = arFS;
+		this.contentType = contentType;
+		this.driveId = driveId;
+		this.entityType = entityType;
+		this.name = name;
+		this.txId = txId;
+		this.unixTime = unixTime;
+	}
+}
+
+export const ENCRYPTED_DATA_PLACEHOLDER = 'ENCRYPTED';
+export type ENCRYPTED_DATA_PLACEHOLDER_TYPE = 'ENCRYPTED';
+
+// A Drive is a logical grouping of folders and files. All folders and files must be part of a drive, and reference the Drive ID.
+// When creating a Drive, a corresponding folder must be created as well. This folder will act as the Drive Root Folder.
+// This seperation of drive and folder entity enables features such as folder view queries.
+export interface ArFSDriveEntity extends ArFSEntity {
+	drivePrivacy: string; // identifies if this drive is public or private (and encrypted)  can only be "public" or "private"
+	rootFolderId: FolderID | ENCRYPTED_DATA_PLACEHOLDER_TYPE; // the uuid of the related drive root folder, stored in the JSON data that is uploaded with each Drive Entity metadata transaction
```

Comment:
Is this necessary here since we added `EncryptedEntityId`?

Reply:
2738016f1886bf5b426e20a42adfbad322765fd6